bdk_chain/src/keychain/keychain_txout_index.rs

```diff
@@ -107,16 +136,80 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-        // TODO: panic if already different descriptor at that keychain
-        self.keychains.insert(keychain, descriptor);
+        let old_descriptor = &*self.keychains.entry(keychain).or_insert(descriptor.clone());
+        assert_eq!(
+            &descriptor, old_descriptor,
+            "keychain already contains a different keychain"
+        );
+    }
+
+    /// Return the lookahead setting for each keychain.
+    ///
+    /// Refer to [`set_lookahead`] for a deeper explanation on `lookahead`.
+    ///
+    /// [`set_lookahead`]: Self::set_lookahead
+    pub fn lookaheads(&self) -> &BTreeMap<K, u32> {
+        &self.lookahead
+    }
+
+    /// Convenience method to call [`set_lookahead`] for all keychains.
+    ///
+    /// [`set_lookahead`]: Self::set_lookahead
+    pub fn set_all_lookaheads(&mut self, lookahead: u32) {
+        let secp = Secp256k1::verification_only();
+        for keychain in &self.keychains.keys().cloned().collect::<Vec<_>>() {
+            self.lookahead.insert(keychain.clone(), lookahead);
+            self.replenish_lookahead(&secp, &keychain);
+        }
+    }
+
+    /// Set the lookahead count for `keychain`.
+    ///
+    /// The lookahead is the number of scripts to cache ahead of the last stored script index. This
+    /// is useful during a scan via [`scan`] or [`scan_txout`].
+    ///
+    /// # Panics
+    ///
+    /// This will panic if `keychain` does not exist.
+    ///
+    /// [`scan`]: Self::scan
+    /// [`scan_txout`]: Self::scan_txout
+    pub fn set_lookahead(&mut self, keychain: &K, lookahead: u32) {
+        self.lookahead.insert(keychain.clone(), lookahead);
+        self.replenish_lookahead(&Secp256k1::verification_only(), keychain);
+    }
+
+    fn replenish_lookahead<C>(&mut self, secp: &Secp256k1<C>, keychain: &K)
+    where
+        C: secp256k1::Verification,
+    {
+        let descriptor = self.keychains.get(keychain).expect("keychain must exist");
+        let next_index = self.last_stored.get(keychain).map_or(0, |v| *v + 1);
+        let lookahead = self.lookahead.get(keychain).map_or(0, |v| *v);
+
+        for index in next_index..next_index + lookahead {
+            let spk = descriptor
+                .at_derivation_index(index)
+                .derived_descriptor(secp)
+                .map(|desc| desc.script_pubkey());
+
+            match spk {
+                Ok(spk) => self
+                    .inner
+                    .insert_script_pubkey((keychain.clone(), index), spk),
+                Err(_) => break,
+            };
+        }
     }
 
     /// Generates script pubkey iterators for every `keychain`. The iterators iterate over all
     /// derivable scripts.
-    pub fn scripts_of_all_keychains(
-        &self,
-    ) -> BTreeMap<K, impl Iterator<Item = (u32, Script)> + Clone> {
+    pub fn all_keychain_scripts(&self) -> BTreeMap<K, impl Iterator<Item = (u32, Script)> + Clone> {
```

Comment:
Let's call the `script_pubkeys` not "scripts".

Reply:
Done